src/validations/html/rules/NodeOutsideBody.js

```diff
@@ -0,0 +1,101 @@
+// This rule checks for any unexpected tags or text that
+// don't get caught by other validators, specifically:
+//
+// * Any tags or text outside the <html> element
+// * Any text that is inside the <html> element but not part
+//   of the <head> or <body> tags
+
+import findLastIndex from 'lodash-es/findLastIndex';
+
+import Code from './Code';
+
+export default class NodeOutsideBody {
+  constructor() {
+    this._openTagStack = [];
+    this._invalidTextLocations = [];
+    this._invalidTagLocations = [];
+  }
+
+  openTag(location, {name}) {
+    if (
+      (this._isTopLevel() && name !== 'html') ||
+      (this._isInsideHtmlTag() && name !== 'head' && name !== 'body')
+    ) {
+      this._invalidTagLocations.push({
+        location,
+        tagName: name,
+      });
+    }
+
+    this._openTagStack.push({location, name});
+  }
+
+  closeTag(location, name) {
+    const openIndex = findLastIndex(
+      this._openTagStack,
+      openTag => openTag.name === name,
+    );
+
+    const isOpened = openIndex >= 0;
+    if (isOpened) {
+      this._openTagStack.splice(openIndex);
+    }
+  }
+
+  text(location, text) {
+    const trimmedText = text.trim();
+    if (trimmedText && (
+      this._isTopLevel() ||
+      this._isInsideHtmlTag() ||
+      this._isDirectChildOfHead()
+    )) {
+      const matchingWhitespaces = text.match(/^(\W+)/gu);
+      let replacementLocation = location;
```

Comment:
Small thing, but I prefer to avoid `let` wherever possible—here I would probably just do an `if`/`else` and `push` the right thing in each branch. Don’t feel strongly about it though.

Reply:
Updated